File: RileyLink/PumpChatViewController.m
//
//  PumpChatViewController.m
//  RileyLink
//
//  Created by Pete Schwamb on 8/8/15.
//  Copyright (c) 2015 Pete Schwamb. All rights reserved.
//

#import "PumpChatViewController.h"
#import "MessageBase.h"
#import "MinimedPacket.h"
#import "NSData+Conversion.h"
#import "Config.h"
#import "RileyLinkBLEManager.h"
#import "HistoryPage.h"
#import "PumpHistoryEventBase.h"
#import "SendAndListenCmd.h"
#import "AppDelegate.h"
#import "PumpOpsSynchronous.h"
#import "PumpOps.h"

@interface PumpChatViewController () {
  IBOutlet UITextView *output;
  IBOutlet UILabel *batteryVoltage;
  IBOutlet UILabel *pumpIdLabel;
}

@property (nonatomic, strong) NSOperationQueue *pumpCommQueue;
@property (nonatomic, strong) PumpOps *pumpOps;

@end

@implementation PumpChatViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  
  pumpIdLabel.text = [NSString stringWithFormat:@"PumpID: %@", [[Config sharedInstance] pumpID]];
  
  AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
  _pumpOps = [[PumpOps alloc] initWithPumpState:appDelegate.pump andDevice:_device];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)addOutputMessage:(NSString*)msg
{
  output.text = [output.text stringByAppendingFormat:@"%@\n", msg];
  NSLog(@"addOutputMessage: %@", msg);
}


- (IBAction)dumpHistoryButtonPressed:(id)sender {
  [_pumpOps getHistoryPage:0 withHandler:^(NSDictionary * _Nonnull res) {
    if (!res[@"error"]) {
      NSData *page = res[@"pageData"];
      NSLog(@"Got page data: %@", [page hexadecimalString]);
      [self decodeHistoryPage:page withModel:res[@"pumpModel"]];
    } else {
      NSString *log = [NSString stringWithFormat:@"Dump of page 0 failed: %@", res[@"error"]];
      [self addOutputMessage:log];
    }
  }];
}

- (void) decodeHistoryPage:(NSData*)data withModel:(NSString*)model {
  
  PumpModel *m = [PumpModel find:model];
  HistoryPage *page = [[HistoryPage alloc] initWithData:data andPumpModel:m];
  
  NSArray *events = [page decode];
  
  for (PumpHistoryEventBase *event in events) {
    [self addOutputMessage:[NSString stringWithFormat:@"Event: %@", event]];
    NSLog(@"Event: %@", event);
  }
}

- (IBAction)pressDownButtonPressed:(id)sender {
  [_pumpOps pressButton];
}

- (IBAction)queryPumpButtonPressed:(id)sender {
  
  [_pumpOps getPumpModel:^(NSString * _Nonnull model) {
    if (model) {
      [self addOutputMessage:[NSString stringWithFormat:@"Pump Model: %@", model]];
    } else {
      [self addOutputMessage:@"Get pump model failed."];
    }
  }];
  
  [_pumpOps getBatteryVoltage:^(NSString * _Nonnull status, float value) {
    [self addOutputMessage:[NSString stringWithFormat:@"Battery Level: %@, %0.02f volts", status, value]];
  }];
}

- (IBAction)tuneButtonPressed:(id)sender {
  [_pumpOps tunePump:^(NSDictionary * _Nonnull res) {
    [self addOutputMessage:[NSString stringWithFormat:@"Tuning results: %@", res]];
  }];
}


@end

File: RileyLink/PumpChatViewController.swift
//
//  PumpChatViewController.swift
//  RileyLink
//
//  Created by Pete Schwamb on 3/12/16.
//  Copyright © 2016 Pete Schwamb. All rights reserved.
//

import UIKit
import MinimedKit
import RileyLinkKit
import RileyLinkBLEKit


class PumpChatViewController: UIViewController {
  
  @IBOutlet var output: UITextView!
  @IBOutlet var batteryVoltage: UILabel!
  @IBOutlet var pumpIdLabel: UILabel!
  
  var pumpOps: PumpOps!
  var device: RileyLinkBLEDevice!

  override func viewDidLoad() {
    super.viewDidLoad()

    pumpIdLabel.text = "PumpID: \(Config.sharedInstance().pumpID ?? "nil")"
    
    let appDelegate = UIApplication.sharedApplication().delegate as! AppDelegate
    pumpOps = PumpOps(pumpState:appDelegate.pump, device:device)
  }

  override func didReceiveMemoryWarning() {
      super.didReceiveMemoryWarning()
      // Dispose of any resources that can be recreated.
  }
  
  func addOutputMessage(msg: String)
  {
    output.text = output.text.stringByAppendingFormat("%@\n", msg)
    NSLog("addOutputMessage: %@", msg)
  }
  
  @IBAction func dumpHistoryButtonPressed(sender: UIButton) {
    let calendar = NSCalendar.currentCalendar()
    let oneDayAgo = calendar.dateByAddingUnit(.Day, value: -1, toDate: NSDate(), options: [])
    pumpOps.getHistoryEventsSinceDate(oneDayAgo!) { (response) -> Void in
      switch response {
      case .Success(let (events, _)):
        for event in events {
          self.addOutputMessage(String(format:"Event: %@", event.dictionaryRepresentation))
          NSLog("Event: %@", event.dictionaryRepresentation)
        }
      case .Failure(let error):
        let errorMsg = String(format:"History fetch failed: %@", String(error))
        self.addOutputMessage(errorMsg)
      }
    }
  }
  
  @IBAction func pressDownButtonPressed(sender: UIButton) {
    pumpOps.pressButton()
  }
  
  @IBAction func queryPumpButtonPressed(sender: UIButton) {
    pumpOps.getPumpModel { (model) -> Void in
      if let model = model {
        self.addOutputMessage(String(format:"Pump Model: %@", model))
      } else {
        self.addOutputMessage("Get pump model failed.")
      }
    }
  
    pumpOps.getBatteryVoltage { (results) -> Void in
      if let results = results {
        self.addOutputMessage(String(format:"Battery Level: %@, %0.02f volts", results.status, results.volts))
      } else {
        self.addOutputMessage("Get battery voltage failed.")        
      }
    }
  }
  
  
  @IBAction func tuneButtonPressed(sender: UIButton) {
    pumpOps.tunePump { (result) -> Void in
      switch result {
      case .Success(let scanResults):
        for trial in scanResults.trials {
          self.addOutputMessage(String(format:"Trial: %0.02f - %d, %0.01f", trial.frequencyMHz, trial.successes, trial.avgRSSI))
        }
        self.addOutputMessage(String(format:"Best Freq: %0.02f", scanResults.bestFrequency))
      case .Failure(let error):
        let errorMsg = String(format:"Tune failed: %@", String(error))
        self.addOutputMessage(errorMsg)
      }
    }
  }
}
